[AMDGPU] Always Inline preserved analyses

jofrn opened the pull request:
When replacing all uses, the structural-hash of the IR can change, so keep track of changes using Changed variable and return it to pass manager.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAlwaysInlinePass.cpp:
```diff
@@ -98,6 +99,7 @@ static bool alwaysInlineImpl(Module &M, bool GlobalOpt) {
       if (TT.getArch() == Triple::amdgcn &&
           A.getLinkage() != GlobalValue::InternalLinkage)
         continue;
+      Preserved = false;
```
The normal way to do this is track changed not preserved

jofrn replied:
ah, thank you!

jofrn committed:
[AMDGPU] Always Inline preserved analyses

When replacing all uses, the structural-hash of the IR can change, so keep track of changes using Changed variable and return it to pass manager.

shiltian reviewed (approved):
LGTM

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAlwaysInlinePass.cpp:
```diff
@@ -166,6 +168,6 @@ ModulePass *llvm::createAMDGPUAlwaysInlinePass(bool GlobalOpt) {
-  alwaysInlineImpl(M, GlobalOpt);
-  return PreservedAnalyses::all();
+  const bool Changed = alwaysInlineImpl(M, GlobalOpt);
+  return Changed ? PreservedAnalyses::none() : PreservedAnalyses::all();
```
This could really preserves most but we should be deleting this pass eventually

jofrn replied:
Good to note this. We can be conservative for now even though I think this may preserve all or most analyses. Perhaps to define a `most` set would be helpful.

Always Inline pass only replaces aliases with functions, and then adds attributes to the function, so the pass could use that set if it stays around. Regardless, there are other passes that could benefit from it as well.

Will submit as it is now. Thank you.
